pkg/live/planner/cluster.go

```diff
@@ -0,0 +1,289 @@
+			actions[index] = handleApplyEvent(e, a)
+		case event.PruneType:
+			id := e.PruneEvent.Identifier
+			index := indexForIdentifier(id, actions)
+			a := actions[index]
+			actions[index] = handlePruneEvent(e, a)
+		case event.DeleteType:
+			id := e.DeleteEvent.Identifier
+			index := indexForIdentifier(id, actions)
+			a := actions[index]
+			actions[index] = handleDeleteEvent(e, a)
+		}
+	}
+	return actions, err
+}
+
+func handleApplyEvent(e event.Event, a planv1alpha1.Action) planv1alpha1.Action {
+	if e.ApplyEvent.Error != nil {
+		a.Type = planv1alpha1.Error
+		a.Error = e.ApplyEvent.Error.Error()
+	} else {
+		switch e.ApplyEvent.Operation {
+		case event.Unchanged:
+			a.Type = planv1alpha1.Skip
+		case event.ServersideApplied:
+			a.After = e.ApplyEvent.Resource
+			if a.Before != nil {
+				// TODO: Unclear if we should diff the full resources here. It doesn't work
+				// well with client-side apply as the managedFields property shows up as
+				// changes. It also means there is a race with controllers that might change
+				// the status of resources.
+				if reflect.DeepEqual(a.Before, a.After) {
+					a.Type = planv1alpha1.Unchanged
+				} else {
+					a.Type = planv1alpha1.Update
+				}
+			} else {
+				a.Type = planv1alpha1.Create
+			}
+		}
+	}
+	return a
+}
+
+func handlePruneEvent(e event.Event, a planv1alpha1.Action) planv1alpha1.Action {
+	if e.PruneEvent.Error != nil {
+		a.Type = planv1alpha1.Error
+		a.Error = e.PruneEvent.Error.Error()
+	} else {
+		switch e.PruneEvent.Operation {
+		case event.Pruned:
+			a.Type = planv1alpha1.Delete
+		case event.PruneSkipped:
+			a.Type = planv1alpha1.Skip
+		}
+	}
+	return a
+}
+
+func handleDeleteEvent(e event.Event, a planv1alpha1.Action) planv1alpha1.Action {
+	if e.DeleteEvent.Error != nil {
+		a.Type = planv1alpha1.Error
+		a.Error = e.DeleteEvent.Error.Error()
+	} else {
+		switch e.DeleteEvent.Operation {
+		case event.Deleted:
+			a.Type = planv1alpha1.Delete
+		case event.DeleteSkipped:
+			a.Type = planv1alpha1.Skip
+		}
+	}
+	return a
+}
+
+func (r *ClusterPlanner) fetchResources(ctx context.Context, e event.Event) ([]planv1alpha1.Action, error) {
+	var actions []planv1alpha1.Action
+	for _, ag := range e.InitEvent.ActionGroups {
+		// We only care about the Apply, Prune and Delete actions.
+		if !(ag.Action == event.ApplyAction || ag.Action == event.PruneAction || ag.Action == event.DeleteAction) {
+			continue
+		}
+		for _, id := range ag.Identifiers {
+			u, _, err := r.resourceFetcher.FetchResource(ctx, id)
+			// If the type doesn't exist in the cluster, then the resource itself doesn't exist.
+			if err != nil && !meta.IsNoMatchError(err) {
+				return nil, err
+			}
+			actions = append(actions, planv1alpha1.Action{
+				Group:     id.GroupKind.Group,
+				Kind:      id.GroupKind.Kind,
+				Name:      id.Name,
+				Namespace: id.Namespace,
+				Before:    u,
+			})
+		}
+	}
+	return actions, nil
+}
+
+type resourceFetcher struct {
+	dc     dynamic.Interface
```

Comment:
Nit: I normally just use `client` or `dynamicClient` (or just embed the interface if it's intrinsic to the type, but that normally ends poorly!).  It just makes it a little easier to read when you come back to it after 6 months vs `dc`

Reply:
Fixed.